### ISSUE TITLE
Push for BasicBertSequenceClassification_basic E2E

### ISSUE THREAD

**Issue by silvasean** [llvm/torch-mlir#342]
@gpetters94 @dan-garvey @cathyzhyi (pls add others)

Creating this bug to centralize task tracking on getting BasicBertSequenceClassification_basic working after https://github.com/llvm/torch-mlir/pull/341 lands. BERT is a milestone model to run -- let's do this!

@cathyzhyi can you post an updated todo op list here? Folks please indicate which ops/bugs you are working on so we don't step on each other's toes :)

**Comment by cathyzhyi** [comment]
Sure! Thanks for opening this issue. Will provide the list for us to collaborate soon.

**Comment by cathyzhyi** [comment]
I summarized the ops that need to be implemented from the [IR output](https://gist.github.com/cathyzhyi/8744f903bc402ae45ea68086f0673d52) into this list. Each one of us could grab the ops and update the status accordingly.

For the status column: 
`-` unimplemented <br>`+` started <br>`X` finished
num | op | status| owner | notes
------------- | ------------- | ------------- | ------------- | -------------
1  | aten::dropout | - ||
2  | aten::embedding | - ||
3  | aten::gelu | - ||
4  | aten::layer_norm | + | cathyzhyi ||
5  | aten::matmul | - ||
6  | aten::softmax.int | - ||
7  | aten::permute | - ||
8  | aten::rsub.Scalar | - ||
9  | aten::slice.Tensor | - || consider aliasing
10 | aten::select.int | - || consider aliasing
11 | aten::view | - || consider aliasing
12 | aten::ones | - || 
13 | aten::size.int | - || might just need a folder
14 | aten::contiguous | - ||
15 | aten::Int.Tensor | - ||
16 | aten::to.dtype | - ||
17 | prim::NumToTensor.Scalar | - ||

**Comment by gpetters94** [comment]
Okay, I'll take `ones`.
 
 num 	op 	status 	owner 	notes
 1 	aten::dropout 	- 		
 2 	aten::embedding 	+ 	cathyzhyi 	
 3 	aten::gelu 	- 		
 4 	aten::layer_norm 	X 	cathyzhyi 	[PR346](https://github.com/llvm/torch-mlir/pull/346)
 5 	aten::matmul 	- 		
 6 	aten::softmax.int 	- 		
 7 	aten::permute 	- 		
 8 	aten::rsub.Scalar 	- 		
 9 	aten::slice.Tensor 	- 		
 10 	aten::select.int 	- 		
 11 	aten::view 	- 		
 12 	aten::ones 	+ 	gpetters94	
 13 	aten::size.int 	+ 	cathyzhyi 	
 14 	aten::contiguous 	- 		
 15 	aten::Int.Tensor 	- 		
 16 	aten::to.dtype 	- 		
 17 	prim::NumToTensor.Scalar 	-

**Comment by silvasean** [comment]
quick note: for softmax, we probably want to decompose it into `softmax(x, dim) => tmp=exp(x); tmp / reduce(tmp, dim, keepdim=True)`
https://pytorch.org/docs/stable/generated/torch.nn.functional.softmax.html

I would recommend adding a pass lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (there are many others like this -- and eventually we will want to automate this, such as by deriving the patterns there from python snippets like `def softmax(x, dim): tmp = torch.exp(x); return torch.reduce(tmp, dim, keepdim=True)`)

**Comment by sogartar** [comment]
Hi, recently joined nod.ai's team. I will implement the GELU activation.

**Comment by cathyzhyi** [comment]
@sogartar @gpetters94 you folks should be able to update the table in https://github.com/llvm/torch-mlir/issues/342#issuecomment-931729772 now.

**Comment by Shukla-Gaurav** [comment]
Hi, I recently joined nod.ai's team. I will take care of `aten.view` op. How can I edit the list?

**Comment by pashu123** [comment]
I will take `aten::matmul` op.

**Comment by cathyzhyi** [comment]
@gprateek93 @gpetters94 @pashu123 don't forget to update the table if your PRs are merged

**Comment by cathyzhyi** [comment]
I am documenting the finds for collaboration. With changes from [repo](https://github.com/cathyzhyi/mlir-npcomp/tree/bert-comp-wip). The model is now hitting a runtime assumption.  The JIT code is [here](https://gist.github.com/cathyzhyi/34df38757a5e6f59d28c52c91c927f6e). And the program jumped to out of line code for aborting at 
```
0x7ffff75b0305: 0f 84 65 02 00 00                             je     0x7ffff75b0570            ; <+9536> jmp here
```

The jump is from inside a loop:
```
0x7ffff75b02d0: 48 8b 74 24 38                                mov    rsi, qword ptr [rsp + 0x38] zy
0x7ffff75b02d5: 48 ff c6                                      inc    rsi
0x7ffff75b02d8: 4c 8b 44 24 78                                mov    r8, qword ptr [rsp + 0x78]
0x7ffff75b02dd: 4c 01 84 24 80 00 00 00                       add    qword ptr [rsp + 0x80], r8 r8 = 0x380 = 896
0x7ffff75b02e5: 48 8b 44 24 50                                mov    rax, qword ptr [rsp + 0x50] = 0xc4
0x7ffff75b02ea: 48 01 44 24 40                                add    qword ptr [rsp + 0x40], rax
0x7ffff75b02ef: 48 01 44 24 48                                add    qword ptr [rsp + 0x48], rax
0x7ffff75b02f4: 4c 01 44 24 08                                add    qword ptr [rsp + 0x8], r8
0x7ffff75b02f9: 48 89 f0                                      mov    rax, rsi
0x7ffff75b02fc: 48 89 74 24 38                                mov    qword ptr [rsp + 0x38], rsi
0x7ffff75b0301: 48 83 fe 0c                                   cmp    rsi, 0xc
0x7ffff75b0305: 0f 84 65 02 00 00                             je     0x7ffff75b0570            ; <+9536> jmp here
0x7ffff75b030b: 4d 85 f6                                      test   r14, r14
0x7ffff75b030e: 7e c0                                         jle    0x7ffff75b02d0            ; <+8864>
```
Still need to figure out the corresponding IR for this loop snippet.

**Comment by cathyzhyi** [comment]
With the [patch1](https://github.com/cathyzhyi/mlir-npcomp/commit/bafcbc1c972d69219b5a4b14f62ab4c10f498258) and [patch2](https://github.com/cathyzhyi/mlir-npcomp/commit/c24e4d21a36591e4e29509fd5e91cff1130f9d47) from [repo](https://github.com/cathyzhyi/mlir-npcomp/tree/bert-comp-wip)  the [MiniLMSequenceClassification_basic](https://github.com/llvm/torch-mlir/blob/main/build_tools/torchscript_e2e_heavydep_tests/minilm_sequence_classification.py) test is passing now!!

Great work everyone!!

**Comment by silvasean** [comment]
Closing this as it's working now!!! Great work everyone :1st_place_medal: This is a huge milestone for the project :)
